libptable/ptable.h

```diff
@@ -0,0 +1,87 @@
+/*
+ * Phoenix-RTOS
+ *
+ * Partition table
+ *
+ * Copyright 2020, 2023 Phoenix Systems
+ * Author: Hubert Buczynski, Lukasz Kosinski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#ifndef _LIB_PTABLE_H_
+#define _LIB_PTABLE_H_
+
+#include <stdint.h>
+
+
+/* Structure of partition table
+ *  _________________________________________________________________________
+ * |       28 B      |                 32 B * n                |     4 B     |
+ * |-----------------|-----------------------------------------|-------------|
+ * | ptable_t header | ptable_part_t 0 | ... | ptable_part_t n | magic bytes |
+ *  -------------------------------------------------------------------------
+ * 
+ *  NOTE: data in partition table should be stored in little endian
+ */
+
+
+/* Partition table magic signature */
+static const uint8_t ptable_magic[] = { 0xde, 0xad, 0xfc, 0xbe };
+
+
+/* Supported partition types */
+enum { ptable_raw = 0x51, ptable_jffs2 = 0x72, ptable_meterfs = 0x75 };
+
+
+typedef struct {
+	uint8_t name[8]; /* Partition name */
+	uint32_t offset; /* Partition offset (in bytes) */
+	uint32_t size;   /* Partition size (in bytes) */
+	uint8_t type;    /* Partition type */
+
+	/* Reserved bytes */
+	union {
+		struct {
+			uint8_t pad[11];
+			uint32_t crc; /* Partition checksum */
+		} __attribute__((packed));
+		uint8_t reserved[15];
+	};
+} __attribute__((packed)) ptable_part_t;
+
+
+typedef struct {
+	uint32_t count; /* Number of partitions */
+
+	/* Reserved bytes */
+	union {
+		struct {
+			uint8_t pad[20];
+			uint32_t crc; /* Header checksum */
+		} __attribute__((packed));
+		uint8_t reserved[24];
+	};
+
+	ptable_part_t parts[0]; /* Partitions */
+} __attribute__((packed)) ptable_t;
```

Comment:
Perhaps it would be better to define structures in this way and add `_Static_asser()`
```suggestion
typedef struct {
	uint8_t name[8]; /* Partition name */
	uint32_t offset; /* Partition offset (in bytes) */
	uint32_t size;   /* Partition size (in bytes) */
	uint8_t type;    /* Partition type */

	/* Reserved bytes */
	uint8_t reserved[11];
#if 1
	uint32_t crc; /* Partition checksum */
#else 
	union {
		uint8_t reserved2[4];
		uint32_t crc; /* Partition checksum */
	} ;
#endif
} ptable_part_t;


typedef struct {
	uint32_t count; /* Number of partitions */

	/* Reserved bytes */
	uint8_t reserved[20];
#if 1
	uint32_t crc; /* Header checksum */
#else
	union {
		uint8_t reserved2[4];
		uint32_t crc; /* Header checksum */
	};
#endif
	ptable_part_t parts[0]; /* Partitions */
} ptable_t;

_Static_assert(sizeof(ptable_part_t == 32, "sizeof(ptable_part_t) != 32");
_Static_assert(sizeof(ptable_t == 28, "sizeof(ptable_t) != 28");
```

Reply:
`_Static_assert` is a C11 feature, we aspire for full C99 support and compatibility.

There is a way:
```
extern char _PTABLE_BUG_ON1_[1 - 2*!!(sizeof(ptable_part_t) != 32)];
extern char _PTABLE_BUG_ON2_[1 - 2*!!(sizeof(ptable_t) != 28)];
```

It's not pretty though, so I'm not sure if we really want it

Reply:
IMHO we could provide our implementation of `_Static_assert` as a part of `libphoenix` as it might me useful regardless of the standard being used (similar constructions were being used before C11, it got into standard just then) - eg. `glibc` provides it for all versions of C standard) if not provided by the compiler already:

![image](https://user-images.githubusercontent.com/3939983/217207950-137a8658-97a6-4957-9bb8-74e7157e8a8e.png)


Not related to this PR, should probably be discussed / resolved separately.

Reply:
Some comments:
1. The use of packed structures increases the size of the code and reduces performance, so you should limit their use to the minimum necessary. IMHO in this case it is not necessary 
2. these structures can be defined without using a union.
3. _Static_assert() I gave as an example, it can be defined in other arbitrary way.